File: client/src/main/java/dev/quasemago/aura/client/infra/searches/jikan/dto/MalSearchResponseDTO.java
package dev.quasemago.aura.client.infra.searches.jikan.dto;

import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.List;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class MalSearchResponseDTO {
    private UserProfile data;

    @Getter
    @Setter
    public static class UserProfile {
        private Integer mal_id;
        @JsonSetter(nulls = Nulls.SKIP)
        private String username = "N/A";
        private String url;
        private Images images;
        @JsonSetter(nulls = Nulls.SKIP)
        private String last_online = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String gender = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String birthday = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String location = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String joined = "N/A";
        private Statistics statistics;
        private Favorites favorites;
        private Updates updates;
        @JsonSetter(nulls = Nulls.SKIP)
        private String about = "N/A";
        private List<ExternalLink> external;
    }

    @Getter
    @Setter
    public static class Images {
        private ImageFormat jpg;
        private ImageFormat webp;
    }

    @Getter
    @Setter
    public static class ImageFormat {
        private String image_url;
    }

    @Getter
    @Setter
    public static class Statistics {
        private AnimeStats anime;
        private MangaStats manga;
    }

    @Getter
    @Setter
    public static class AnimeStats {
        private Double days_watched;
        private Double mean_score;
        private Integer watching;
        private Integer completed;
        private Integer on_hold;
        private Integer dropped;
        private Integer plan_to_watch;
        private Integer total_entries;
        private Integer rewatched;
        private Integer episodes_watched;
    }

    @Getter
    @Setter
    public static class MangaStats {
        private Double days_read;
        private Double mean_score;
        private Integer reading;
        private Integer completed;
        private Integer on_hold;
        private Integer dropped;
        private Integer plan_to_read;
        private Integer total_entries;
        private Integer reread;
        private Integer chapters_read;
        private Integer volumes_read;
    }

    @Getter
    @Setter
    public static class Favorites {
        private List<FavoriteAnime> anime;
        private List<FavoriteManga> manga;
        private List<FavoriteCharacter> characters;
        private List<FavoritePeople> people;
    }

    @Getter
    @Setter
    public static class FavoriteAnime {
        private Integer mal_id;
        private String url;
        private Images images;
        @JsonSetter(nulls = Nulls.SKIP)
        private String title = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String type = "N/A";
        private Integer start_year;
    }

    @Getter
    @Setter
    public static class FavoriteManga {
        private Integer mal_id;
        private String url;
        private Images images;
        @JsonSetter(nulls = Nulls.SKIP)
        private String title = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String type = "N/A";
        private Integer start_year;
    }

    @Getter
    @Setter
    public static class FavoriteCharacter {
        private Integer mal_id;
        private String url;
        private Images images;
        @JsonSetter(nulls = Nulls.SKIP)
        private String name = "N/A";
    }

    @Getter
    @Setter
    public static class FavoritePeople {
        private Integer mal_id;
        private String url;
        private Images images;
        @JsonSetter(nulls = Nulls.SKIP)
        private String name = "N/A";
    }

    @Getter
    @Setter
    public static class Updates {
        private List<AnimeUpdate> anime;
        private List<MangaUpdate> manga;
    }

    @Getter
    @Setter
    public static class AnimeUpdate {
        private AnimeEntry entry;
        private Double score;
        @JsonSetter(nulls = Nulls.SKIP)
        private String status = "N/A";
        private Integer episodes_seen;
        private Integer episodes_total;
        @JsonSetter(nulls = Nulls.SKIP)
        private String date = "N/A";
    }

    @Getter
    @Setter
    public static class MangaUpdate {
        private MangaEntry entry;
        private Double score;
        @JsonSetter(nulls = Nulls.SKIP)
        private String status = "N/A";
        private Integer chapters_read;
        private Integer chapters_total;
        @JsonSetter(nulls = Nulls.SKIP)
        private String date = "N/A";
    }

    @Getter
    @Setter
    public static class AnimeEntry {
        private Integer mal_id;
        private String url;
        private Images images;
        @JsonSetter(nulls = Nulls.SKIP)
        private String title = "N/A";
    }

    @Getter
    @Setter
    public static class MangaEntry {
        private Integer mal_id;
        private String url;
        private Images images;
        @JsonSetter(nulls = Nulls.SKIP)
        private String title = "N/A";
    }

    @Getter
    @Setter
    public static class ExternalLink {
        @JsonSetter(nulls = Nulls.SKIP)
        private String name = "N/A";
        private String url;
    }
}

File: src/infrastructure/jikan/dtos/mal-search-response-dto.ts
import { IsString, IsInt, IsOptional, IsNumber, IsArray, ValidateNested } from "class-validator";
import { Type } from "class-transformer";

export class MalSearchResponseDTO {
  @ValidateNested()
  @Type(() => UserProfile)
  data?: UserProfile;
}

export class UserProfile {
  @IsInt()
  mal_id!: number;

  @IsString()
  @IsOptional()
  username: string = "N/A";

  @IsString()
  url?: string;

  @ValidateNested()
  @Type(() => Images)
  images?: Images;

  @IsString()
  @IsOptional()
  last_online: string = "N/A";

  @IsString()
  @IsOptional()
  gender: string = "N/A";

  @IsString()
  @IsOptional()
  birthday: string = "N/A";

  @IsString()
  @IsOptional()
  location: string = "N/A";

  @IsString()
  @IsOptional()
  joined: string = "N/A";

  @ValidateNested()
  @Type(() => Statistics)
  statistics?: Statistics;

  @ValidateNested()
  @Type(() => Favorites)
  favorites?: Favorites;

  @ValidateNested()
  @Type(() => Updates)
  updates?: Updates;

  @IsString()
  @IsOptional()
  about: string = "N/A";

  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => ExternalLink)
  external?: ExternalLink[];
}

export class Images {
  @ValidateNested()
  @Type(() => ImageFormat)
  jpg?: ImageFormat;

  @ValidateNested()
  @Type(() => ImageFormat)
  webp?: ImageFormat;
}

export class ImageFormat {
  @IsString()
  image_url?: string;
}

export class Statistics {
  @ValidateNested()
  @Type(() => AnimeStats)
  anime?: AnimeStats;

  @ValidateNested()
  @Type(() => MangaStats)
  manga?: MangaStats;
}

export class AnimeStats {
  @IsNumber() days_watched?: number;
  @IsNumber() mean_score?: number;
  @IsInt() watching?: number;
  @IsInt() completed?: number;
  @IsInt() on_hold?: number;
  @IsInt() dropped?: number;
  @IsInt() plan_to_watch?: number;
  @IsInt() total_entries?: number;
  @IsInt() rewatched?: number;
  @IsInt() episodes_watched?: number;
}

export class MangaStats {
  @IsNumber() days_read?: number;
  @IsNumber() mean_score?: number;
  @IsInt() reading?: number;
  @IsInt() completed?: number;
  @IsInt() on_hold?: number;
  @IsInt() dropped?: number;
  @IsInt() plan_to_read?: number;
  @IsInt() total_entries?: number;
  @IsInt() reread?: number;
  @IsInt() chapters_read?: number;
  @IsInt() volumes_read?: number;
}

export class Favorites {
  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => FavoriteAnime)
  anime?: FavoriteAnime[];

  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => FavoriteManga)
  manga?: FavoriteManga[];

  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => FavoriteCharacter)
  characters?: FavoriteCharacter[];

  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => FavoritePeople)
  people?: FavoritePeople[];
}

export class FavoriteAnime {
  @IsInt() mal_id?: number;
  @IsString() url?: string;

  @ValidateNested()
  @Type(() => Images)
  images?: Images;

  @IsString()
  @IsOptional()
  title: string = "N/A";

  @IsString()
  @IsOptional()
  type: string = "N/A";

  @IsInt()
  @IsOptional()
  start_year?: number;
}

export class FavoriteManga {
  @IsInt() mal_id?: number;
  @IsString() url?: string;

  @ValidateNested()
  @Type(() => Images)
  images?: Images;

  @IsString()
  @IsOptional()
  title: string = "N/A";

  @IsString()
  @IsOptional()
  type: string = "N/A";

  @IsInt()
  @IsOptional()
  start_year?: number;
}

export class FavoriteCharacter {
  @IsInt() mal_id?: number;
  @IsString() url?: string;

  @ValidateNested()
  @Type(() => Images)
  images?: Images;

  @IsString()
  @IsOptional()
  name: string = "N/A";
}

export class FavoritePeople {
  @IsInt() mal_id?: number;
  @IsString() url?: string;

  @ValidateNested()
  @Type(() => Images)
  images?: Images;

  @IsString()
  @IsOptional()
  name: string = "N/A";
}

export class Updates {
  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => AnimeUpdate)
  anime?: AnimeUpdate[];

  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => MangaUpdate)
  manga?: MangaUpdate[];
}

export class AnimeUpdate {
  @ValidateNested()
  @Type(() => AnimeEntry)
  entry?: AnimeEntry;

  @IsNumber()
  score?: number;

  @IsString()
  @IsOptional()
  status: string = "N/A";

  @IsInt()
  episodes_seen?: number;

  @IsInt()
  episodes_total?: number;

  @IsString()
  @IsOptional()
  date: string = "N/A";
}

export class MangaUpdate {
  @ValidateNested()
  @Type(() => MangaEntry)
  entry?: MangaEntry;

  @IsNumber()
  score?: number;

  @IsString()
  @IsOptional()
  status: string = "N/A";

  @IsInt()
  chapters_read?: number;

  @IsInt()
  chapters_total?: number;

  @IsString()
  @IsOptional()
  date: string = "N/A";
}

export class AnimeEntry {
  @IsInt() mal_id?: number;
  @IsString() url?: string;

  @ValidateNested()
  @Type(() => Images)
  images?: Images;

  @IsString()
  @IsOptional()
  title: string = "N/A";
}

export class MangaEntry {
  @IsInt() mal_id?: number;
  @IsString() url?: string;

  @ValidateNested()
  @Type(() => Images)
  images?: Images;

  @IsString()
  @IsOptional()
  title: string = "N/A";
}

export class ExternalLink {
  @IsString()
  @IsOptional()
  name: string = "N/A";

  @IsString()
  url?: string;
}
